README.md
# FootballBot
A Football Simulation Bot that lets you play in your own world with your own team and manage it as a coach! You will be able to play in tournaments, buy players, sell players, and much more!

This README will be updated along with the commits, explaining the code's intended functions.

## Player Generation
The code for player generation resides in `./Backend/generation.py`, while the text files with data reside in `./Data/Generation/` which is used to randomly create the player names. The code generates a player name, assigns it a rating and an age along with its position of play. 

The generation script is to be activated when a user creates a new world, thus needing new players. It returns a list of Player class objects which have attributes that can be accessed. (The class can be printed as well to display relevant information)

Each player is assigned to a team, and all the relevant databases are created for the world. The data is then passed into the database and saved. All of this is done when the user types `/register` after which a new world is created (only the user can play in it, although multiplayer will be developed in the future). To unregister, just type `/unregister` and all the data that was saved will be removed (IRREVERSIBLE).

## Player Value Calculation

The calculation of the player's value (transfer value) is done based on which age category the player's age falls into:

1. Under 20
2. Under 25
3. Under 30
4. Above 30

This is to deal with complex stuff like potential, since a higher rated young player is worth more than an older one because the player has more time to improve and increase their rating. The value also depends on the player's position, since a GK is generally less worth than a ST in real life.

## Player card generation

For showcasing different aspects such as position and rating of a particular player, the `/view <field> <name>` command can be used to view a digitally created card like the one that follows:

![image](https://user-images.githubusercontent.com/75207653/201659866-978d9b20-e5ec-44b9-976f-e077977196ed.png)

NOTE: `<field>` basically has to be either `team` or `player` for now, as this is the same command that is used to obtain the squad of any particular team.

## Mail System

The Mail System is a way for the user to receive messages about the events of their world, for example a note from the team board. Note that it can also be used by teams to send contracts to users which can be accepted or rejected by the user in the interface itself.


A mail is divided based on its type: 

1. Message
2. Offer

## Starting XI 

The logic for finding a team's starting XI is simple: Find the best players for each position. For now, this is the maximum complexity. the `starting11` function returns a dictionary with all the starting players included. A team's starting XI can be viewed with the command `/view 11 <team>` (NOTE: Proportions will be fixed later)

![image](https://user-images.githubusercontent.com/119870649/206257976-d61168d7-1ad8-477f-b75f-e40266c3a210.png)

## Currently Working On

Game play logic (without weights) and interaction with user with chance generation and more.


CHANGELOG.md
# Change Log
#### Made by Rayan  
<br />
<br />

## Version 0.0.1
___

- Made function generates individual players

## Version 0.0.2

- Made algorithm that assigns players to teams
- Made ```register``` and ```unregister``` functions that add user's id and name to a database (or removes it)

## Version 0.0.3

- Created `create_world` function that creates a new world along with the necessary databases and commits all data (to be imported and called to create new world, and supplied userid)
- Made algorithm to calculate player value, for transfer purposes
- Connected `register` and `unregister` functions to the generator code, creating or destroying the user's world

## Version 0.0.4

- Remade `world` DB's structure for easier usage and better efficiency
- Decreased high rated player generation
- Fixed problem involving very high rated youngsters


## Version 0.0.5

- Created functions for player card image generation, which can be viewed with the command `show`, to search for a specific player's card
- Made a team's squad viewable

## Version 0.0.6

- Added a mail system for the user to be able to interact with the world
  
## Version 0.0.7

- Created logic for finding a team's starting XI
- Made team starting XI viewable using +show 11 <team> command, which gives a generated image of the starting team (best players)
  
## Version 0.0.8

- Migrated to slash commands

### Work to be done

- [x] Register User
- [x] Generate World
- [x] Calculate player value
- [x] Make player GUI to view players
- [x] Make team GUI to view teams 
- [x] Create mail inbox
- [x] Create AI for finding Starting XI
- [x] Migrate to slash commands 
- [ ] Fix some proportions
- [ ] Generate offers for user
- [ ] League Fixture Generation
- [ ] View league standings
- [ ] Match Simulation
- [ ] Match Gameplay Logic (currently working on)
- [ ] Gameplay Weighting
- [ ] Add all the Prem teams

Backend/generator.py
import random, os, math
from easy_sqlite3 import *

WEIGHTS = [i for i in range(105, 70, -3)] + [i for i in range(70, 40, -2)] + [i for i in range(40, 1, -5)]

class Player:
    def __init__(self, name, pos, rating, age=None):
        self.name = name
        self.pos = pos
        self.rating = rating
        self.age = age if age else self.gen_age()
    
    def gen_age(self):
        ages =  [i for i in range(17, 34)]
        if self.rating >= 90:
            return random.choices(ages, k=1, weights=[i for i in range(1, 5)] + [i for i in range(5, 16, 2)] + [i for i in range(16, 26, 3)] + [25, 22, 19])[0]
        elif self.rating >= 85:
            return random.choices(ages, k=1, weights=[i for i in range(5, 9)] + [i for i in range(9, 20, 2)] + [i for i in range(20, 30, 3)] + [30, 27, 24])[0]
        elif self.rating >= 75:
            return random.choice(ages)
        else:
            return random.choices(ages, k=1, weights=[i for i in range(34, 17, -1)])[0]


    def __repr__(self):
        return f"{self.name} - {self.pos} - ({self.rating}) -- {self.age} yrs"

    
def generate_player(name, pos):
    rating = random.choices([i for i in range(60, 95)], k=1, weights=WEIGHTS)[0]
    return Player(name, pos, rating)

def generate_players():
    POSITIONS = ['GK', 'LB', 'CB', 'CB', 'RB', 'LM', 'CM', 'RM', 'LW', 'ST', 'RW']
    players = {}

    with open('./Data/Generation/first_names.txt', 'r') as f:
        first_names = f.readlines()

    with open('./Data/Generation/last_names.txt', 'r') as f:
        last_names = f.readlines()
    
    for _ in range(7):
        for pos in POSITIONS:
            for _ in range(3):
                name = f"{random.choice(first_names).title()} {random.choice(last_names).title()}".replace('\n', '')
                players[pos] = players.get(pos, []) + [generate_player(name, pos)]

    return players

def generate_teams():
    TEAMS = ["Liverpool", "Manchester City", "Manchester United", "Aston Villa", "Arsenal", "Tottenham"]

    players = generate_players()

    teams = {}

    for li in players.values():
        random.shuffle(li)

    for team in TEAMS:
        teams[team] = []
        for _ in range(3):
            for pos in players:
                teams[team].append(players[pos].pop())
    
    left = [p for player in players.values() for p in player]
    
    while left:
        teams[random.choice(TEAMS)].append(left.pop())

    return teams
 
def create_tables(teams, user):
    world_db = Database(f"./Data/Worlds/{user}/world", spaces=False)
    league_db = Database(f"./Data/Worlds/{user}/league")
    player_db = Database(f"./Data/Worlds/{user}/players")
    team_db = Database(f'./Data/Worlds/{user}/teams', spaces=False)

    league_db.create_table('league', {'name':TEXT, 'matches':INT, 'gd':INT, 'scored':INT, 'conceded':INT,
                    'wins':INT, 'draws':INT, 'losses':INT, 'points':INT})
    league_db.create_table('player_stats', {'player':TEXT, 'team':TEXT, 'goals':INT, 'assists':INT})

    world_db.create_table("teams", {'team':TEXT, 'coach':TEXT, 'players':INT, 'teamrating':INT, 'transferbudget':INT})

    for team in teams:
        team_db.create_table(team, {'player':TEXT, 'team':TEXT, 'position':TEXT, 'goals':INT, 'assists':INT, 'value':INT,
                    'rating':INT, 'age':INT, 'happiness':INT})


    player_db.create_table("players", {'player':TEXT, 'team':TEXT, 'position':TEXT, 'goals':INT, 'assists':INT, 'value':INT,
                    'rating':INT, 'age':INT, 'happiness':INT})

    del world_db, league_db, player_db, team_db



    mail_db = Database(f"./Data/Worlds/{user}/mail")
    mail_db.create_table('inbox', {'message':STR, 'subject':STR, 'sender':STR, 'type':STR, 'status':INT})
    mail_db.create_table('outbox', {'message':STR, 'subject':STR, 'receiver':STR, 'status':INT})

    del mail_db

def calculate_value(player):
    
    pos = {'GK':2500000, 'LB':10000000, 'CB':7500000, 'LM':15000000, 'CM':15000000,
                 'RM':15000000, 'LW':20000000, 'ST':22500000, 'RW':20000000, 'RB':10000000}

    value = int(pos[player.pos] + math.exp(player.rating/8) * 935 + random.randint(100, 500) * 250 - player.age * 100_000)

    if player.age < 20:
        return value

    if player.age < 25:
        sub = player.age * 75_000
        return value - (sub if value > sub else value//2)

    elif player.age < 30:
        sub = player.age * 200_000
        return value - (sub if value > sub else value//4)
    else:
        value = int(pos[player.pos] + math.exp(player.rating/4.5) * 0.05 + random.randint(10, 50) * 250 - player.age * 750_000)
        return  value if value > 50_000 else int(math.exp(player.rating/4.9) * 0.2 + random.randint(10, 50) * 250)


def average_rating(players):
    total = sum([p.rating for p in players])
    return int(total / len(players))

def add_players(teams, user):
    team_db = Database(f"./Data/Worlds/{user}/teams", spaces=False)
    world_db = Database(f"./Data/Worlds/{user}/world", spaces=False)
    player_db = Database(f"./Data/Worlds/{user}/players")

    for db in (team_db, world_db, player_db):
        db.initiate_transaction()

    for team, players in teams.items():
        world_db.insert("teams", (team, 'Bot', len(players), average_rating(players), 100000))
        for p in players:
            player = (p.name, team, p.pos, 0, 0, calculate_value(p), p.rating, p.age, 10)
            team_db.insert(team, player)
            player_db.insert("players", player)

    for db in (team_db, world_db, player_db):
        db.commit()

    del team_db, world_db, player_db


def generate_world(user):
    #Creating a folder
    base_path = os.getcwd()
    teams = generate_teams()
    try:
        folder_path = f"{base_path}\\Data\\Worlds\\{user}"
        os.mkdir(folder_path)

    except FileExistsError:
        pass
    
    create_tables(teams, user)
    add_players(teams, user)

def destroy_world(user):
    from shutil import rmtree
    folder_path = f"{os.getcwd()}\\Data\\Worlds\\{user}"

    rmtree(folder_path)
